Add ChoicePrompt

enzocanoo opened the pull request:
* Implement ChoicePrompt using ChoiceRecognizer.
* Add ChoicePrompt tests.
* Modify ConfirmPrompt to use BooleanRecognizer.

Stevenic commented:
Do we honestly expect developers to have to build a model for every possible list of choices they want to recognize?  This is not at parity at all with the choice library in the JavaScript SDK and to be honest it looks complicated as hell.

Stevenic commented on tests/Microsoft.Bot.Builder.Prompts.Tests/ChoicePromptTests.cs:
```diff
@@ -0,0 +1,169 @@
+            }
+        }
+    }
+}
```
Wouldn't ['yellow', 'blue', 'red'] be a simpler way to express a list of choices?  Even expanding to include a list of synonyms shouldn't be more complicated then passing in an array of `Choice` objects.

enzocanoo commented:
We updated the `ChoicePrompt` constructor to receive a dictionary with a list of synonyms to match (or a regex) and a result value for each item of the list. Also, the constructor now receive the recognizer's params (`allowPartialMatch` and `maxDistance`)

``` C#
var dictionaryColors = new Dictionary<IEnumerable<string>, Color>()
{
  {
    new string[] { "amber", "aureolin", "gold", "mustard", "vanilla", "yellow" },
    Color.Yellow
  },
  {
    new string[] { "azure", "blue", "blue-gray", "cyan", "indigo", "turquoise" },
    Color.Blue
  },
  {
    new string[] { "amaranth", "auburn", "magenta", "raspberry", "rose", "ruby" },
    Color.Red
  }
};
var testPrompt = new ChoicePrompt<Color>(Culture.English, dictionaryColors, allowPartialMatch: true, maxDistance: 1);
```

However, if the user wants to create their own `CustomChoiceModel`, we added a new `ChoiceModelBuilder` class that generates a `ChoiceModel` using the same approach of the dictionary with the match values and the corresponding value result.

``` C#
var model = new ChoiceModelBuilder<Color>(dictionaryColors)
  .WithCulture(Culture.English)
  .WithAllowPartialMatch(true)
  .WithMaxDistance(1)
  .Build();

var prompt = new ChoicePrompt<Color>(model);
```

Stevenic commented:
This is better but we're still missing things like the `ChoiceFactory` class for presenting the list of choices to the user in a number of ways.

johnataylor commented:
This looks like a nice implementation however its really quite different to the JavaScript. My feeling is that the high order bit here is to make things the same to the degree that we can. Let's hold off on merging this right now and we can see if a deeper consistency is where we want to be.

enzocanoo commented:
@Stevenic, @johnataylor, we implemented static methods from `ChoiceFactory` as extension methods of `IEnumerable<Choice>`. Also, we implemented the `ChoicePrompt` dialog from the Microsoft.Bot.Builder.Dialogs project.

cleemullins commented:
This PR is deprecated with the work @johnataylor has done.
